azure-communication-ui/calling/src/main/java/com/azure/android/communication/ui/calling/data/CallHistoryRepository.kt

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.android.communication.ui.calling.data
+
+import android.content.ContentValues
+import android.content.Context
+import android.database.sqlite.SQLiteDatabase
+import com.azure.android.communication.ui.calling.data.model.CallHistoryRecordData
+import com.azure.android.communication.ui.calling.logger.Logger
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+import org.threeten.bp.Instant
+import org.threeten.bp.OffsetDateTime
+import org.threeten.bp.ZoneId
+
+internal interface CallHistoryRepository {
+    suspend fun insert(callId: String, callDateTime: OffsetDateTime)
+    suspend fun getAll(): List<CallHistoryRecordData>
+}
+
+internal class CallHistoryRepositoryImpl(
+    private val context: Context,
+    private val logger: Logger
+) : CallHistoryRepository {
+
+    override suspend fun insert(callId: String, callDateTime: OffsetDateTime) {
+        return withContext(Dispatchers.IO) {
+            // SQLite does not allow concurrent writes. Need to queue them via lock.
+            synchronized(dbAccessLock) {
+                // Using a new db instance instead of caching one as we do not have a
+                // reliable event when to dispose it.
+                DbHelper(context).writableDatabase
+                    .use { db ->
+                        val values = ContentValues().apply {
+                            put(CallHistoryContract.COLUMN_NAME_CALL_ID, callId)
+                            put(CallHistoryContract.COLUMN_NAME_CALL_DATE, callDateTime.toInstant().toEpochMilli())
+                        }
+
+                        val result = db.insert(CallHistoryContract.TABLE_NAME, null, values)
+                        if (result == -1L) {
+                            logger.warning("Failed to save call history record.")
+                        }
+
+                        // Execute cleanup separately (not in one transaction) in case of it fails,
+                        // so it does not affect insert.
+                        cleanupOldRecords(db)
+                    }
+            }
+        }
+    }
+
+    override suspend fun getAll(): List<CallHistoryRecordData> {
+        return withContext(Dispatchers.IO) {
+            // Using a new db instance instead of caching one as we do not have a
+            // reliable event when to dispose it.
+            DbHelper(context).writableDatabase.use { db ->
+                cleanupOldRecords(db)
```

Comment:
This needs to be in a synchronized block (for the same reason you have one for inserts). A delete is also a "write" operation, so same constraints apply.

I would just remove this line though instead of bumping a simple read to be a write transaction. It's not strictly necessary to cleanup here (you're already cleaning up on inserts).

Reply:
Fixed.